Makefile

```diff
@@ -21,7 +23,10 @@ PACKAGES := $(shell go list $(BUILDFLAGS) ./... | grep -v github.com/containers/
 # (and the user will probably find out because the cgo compilation will fail).
 GPGME_ENV = CGO_CFLAGS="$(shell gpgme-config --cflags 2>/dev/null)" CGO_LDFLAGS="$(shell gpgme-config --libs 2>/dev/null)"
 
-all: tools .gitvalidation test validate
+all: tools build test validate .gitvalidation
+
+build: vendor
+	@$(GPGME_ENV) go build $(BUILDFLAGS) $(PACKAGES)
```

Comment:
`go build --help`:
> When compiling multiple packages or a single non-main package,
> build compiles the packages but discards the resulting object,
> serving only as a check that the packages can be built.

What does this do in addition to the existing `test` target?

The `test` target implicitly checks that each package with tests can be built, and AFAICS even for packages with no `*_test.go` files it seems to do a full compilation, or something very close. Is there any extra benefit to a separate `go build`?

Reply:
> What does this do in addition to the existing test target

It does less in this regards. It just complies. I find it extremely odd to have a Makefile which does not have a build/compile target. The _test_ target might do this implicitly, but imo this is not the same. If I just want to check whether something complies, it feels unnatural to call test. Also as you say, I am also wondering what the exact behaviour is of go test for packages w/o test files. Is it actually specified that it does a full compile as well?

Reply:
> I find it extremely odd to have a Makefile which does not have a build/compile target.

\*shrug\* I find Golang’s lack of shared libraries problematic for many reasons, but that’s Golang for you. The point is that `go build` does not create any artifact, so a
```make
build:
    /bin/true
```
rule would work just as well, and be equally pointless.

>  If I just want to check whether something complies, it feels unnatural to call test.

If the rule is not producing anything consumable, and it’s only effect is verifying whether the code is usable, that seems entirely appropriate for the “run self-tests of the package” target, i.e. `test`.

Reply:
> Is it actually specified that it does a full compile as well?

`go help test`:
```
'Go test' recompiles each package along with any files with names matching
the file pattern "*_test.go".
```

That seems to specify that the package is always compiled, though it isn’t as explicit as we might like.

Reply:
> The point is that go build does not create any artifact, so a

> build:
>    /bin/true
> rule would work just as well, and be equally pointless.

Why is something which verifies that all packages compile pointless? Why does the outcome need to be a binary? 

Since we are quoting from the go help: `go build -f`:

> When compiling multiple packages or a single non-main package,
> build compiles the packages but discards the resulting object,
> serving only as a check that the packages can be built.

which is exactly what I want. It does explicitly what I want, whereas _go test_ does it as a side-effect. Often I just want to know whether something compiles. Running tests is not needed or overhead, forcing me to abort the command.

> That seems to specify that the package is always compiled, though it isn’t as explicit as we might like.

Right, hence `go build`

Reply:
You started with
> I find it extremely odd to have a Makefile which does not have a build/compile target.

Well, this is a build/compile target which does produce anything built/compiled.

Reply:
> Often I just want to know whether something compiles. Running tests is not needed or overhead

I don’t quite get what that is useful for. I guess my development process is different; my IDE immediately points out syntax errors in the current package, making a `Makefile` rule to compile all packages unnecessary.

---

Anyway, my primary concern is that having both `build` and `test` in `all` means that in a edit-compile-test cycle everything is unnecessarily compiled twice, slowing progress down.

I guess I could live with a `build` target which is not in `all`, though I would still find it unnecessary and pretty confusing.

Reply:
> I guess I could live with a build target which is not in all,

Ok. Fair enough.

>  though I would still find it unnecessary and pretty confusing.

I think we have to agree to disagree ;-) I find having to use 'make test' counter-intuitive.